lib/util/exec/common.ts

```diff
@@ -1,8 +1,143 @@
+  opts: RawExecOptions & { maxBuffer: number }
+): [Buffer[], Buffer[]] {
+  const stdout: Buffer[] = [];
+  const stderr: Buffer[] = [];
+  let stdoutLen = 0;
+  let stderrLen = 0;
+
+  cp.stdout?.on('data', (chunk: Buffer) => {
+    // process.stdout.write(data.toString());
+    const len = Buffer.byteLength(chunk, encoding);
+    stdoutLen += len;
+    if (stdoutLen > opts.maxBuffer) {
+      cp.emit('error', new Error('stdout maxBuffer exceeded'));
+    } else {
+      stdout.push(chunk);
+    }
+  });
+
+  cp.stderr?.on('data', (chunk: Buffer) => {
+    // process.stderr.write(data.toString());
+    const len = Buffer.byteLength(chunk, encoding);
+    stderrLen += len;
+    if (stderrLen > opts.maxBuffer) {
+      cp.emit('error', new Error('stderr maxBuffer exceeded'));
+    } else {
+      stderr.push(chunk);
+    }
+  });
+  return [stdout, stderr];
+}
+
+export function exec(cmd: string, opts: RawExecOptions): Promise<ExecResult> {
+  return new Promise((resolve, reject) => {
+    const maxBuffer = opts.maxBuffer ?? 10 * 1024 * 1024; // Set default max buffer size to 10MB
+    const cp = spawn(cmd, {
+      ...opts,
+      // force detached on non WIN platforms
+      // https://github.com/nodejs/node/issues/21825#issuecomment-611328888
+      detached: process.platform !== 'win32',
+      shell: typeof opts.shell === 'string' ? opts.shell : true, // force shell
+    });
+
+    // handle streams
+    const [stdout, stderr] = initStreamListeners(cp, {
+      ...opts,
+      maxBuffer,
+    });
+
+    // handle process events
+    cp.on('error', (error) => {
+      kill(cp, 'SIGTERM');
+      reject(new ExecError(error.message, rejectInfo(), error));
+    });
+
+    cp.on('exit', (code: number, signal: NodeJS.Signals) => {
+      if (NONTERM.includes(signal)) {
+        return;
+      }
+
+      if (signal) {
+        const message = `Process signaled with "${signal}"`;
+        kill(cp, signal);
+        reject(new ExecError(message, { ...rejectInfo(), signal }));
+        return;
+      }
+      if (code !== 0) {
+        const message = `Process exited with exit code "${code}"`;
+        reject(new ExecError(message, { ...rejectInfo(), exitCode: code }));
+        return;
+      }
+      resolve({
+        stderr: stringify(stderr),
+        stdout: stringify(stdout),
+      });
+    });
+
+    function rejectInfo(): ExecErrorData {
+      return {
+        cmd: cp.spawnargs.join(' '),
+        options: opts,
+        stdout: stringify(stdout),
+        stderr: stringify(stderr),
+      };
+    }
+  });
+}
+
+function kill(cp: ChildProcess, signal: NodeJS.Signals): boolean {
+  try {
+    /**
+     * If `pid` is negative, but not `-1`, signal shall be sent to all processes
+     * (excluding an unspecified set of system processes),
+     * whose process group ID (pgid) is equal to the absolute value of pid,
+     * and for which the process has permission to send a signal.
+     */
+    // process.kill(-(cp.pid as number), signal);
```

Comment:
?? So this didn't solve the root issue? Should be removed when not used.

Reply:
it does see [this](https://github.com/renovatebot/renovate/pull/16414#issuecomment-1179670707)

this is a refactor only. once this get merged we will uncomment this and enable the pid range hack.

Reply:
OK, lets add an additional comment, so this doesn't get silently removed by someone on cleanup 😉